### PR TITLE
Enhance site position avatar UI/UX for RFC-001

### LINKED ISSUE #268: [Tracking] RFC-001: Allow guests to collaborate on multiple remote buffers
We intend to implement [RFC-001](https://github.com/atom/teletype/blob/master/doc/rfcs/001-allow-guests-to-open-multiple-remote-buffers.md) via a series of pull requests. This will allow developers to start benefiting from the completed RFC-001 functionality without having to wait for _all_ of RFC-001 to be completed. We'll use this issue to track progress.

## TODO

- [x] When following the host, any time they switch to a new buffer, a new editor for that remote buffer is automatically added to your workspace and focused. Existing editors for previous remote buffers are not automatically closed when this switch occurs. (#262) ([v0.3.0](https://github.com/atom/teletype/releases/tag/v0.3.0))
- [x] When a host closes a buffer, it is removed from all guest portals. (#262) ([v0.3.0](https://github.com/atom/teletype/releases/tag/v0.3.0))
- [x] When you are following a guest and the guest navigates to a different buffer in the same portal, you follow them to that buffer. (#262) ([v0.3.0](https://github.com/atom/teletype/releases/tag/v0.3.0))
- [ ] Show site position avatars for participants that are in non-portal pane items (#332)
- [x] Allow guests to use fuzzy-finder to open any remote editor shared by the host (https://github.com/atom/teletype/pull/323)
- [x] Once the fuzzy-finder support is in place, consider removing the logic that causes guests to automatically re-follow the host when the portal goes from *n* to zero remote items. (If the guest is working in a local editor when the portal goes from *n* to zero remote items, and we automatically re-follow the host, it will yank the guest away from their current editor. We'd like to avoid that. 😬) (https://github.com/atom/teletype/pull/323)
- [x] Close remote buffers on guest when portal gets closed or disconnected (https://github.com/atom/teletype/pull/319)
---

Refs: https://github.com/atom/teletype/pull/231 (which introduced RFC-001)

### LINKED ISSUE #286: Distinguishing between users [from question in RFC-001]
Some ways I can think of to distinguish between people are size and saturation.
You could add a grey overlay to contributors outside of portal. 
You could make all contributors outside the current file smaller and in a fourth row.
like this (I also included my enchancement for tab icons #275 ):
teletype

(I didn't wan't to clutter #268 so I made my own issue with advice)

### PR DESCRIPTION
**Pull request by jasonrudolph**
Depends on https://github.com/atom/teletype-client/pull/54

### Description of the Change

This pull request updates the site position avatars in support of #268.

#286 explored various designs, and this pull request implements the design described in the last mockup in https://github.com/atom/teletype/issues/286#issuecomment-352384609:

mockup

### Verification Process

- [x] Verify proper rendering of avatars and link icon in popular themes
- [x] Verify proper rendering of link icon when using small and large font sizes
- [x] Given a participant that is viewing the same editor that you're viewing, verify that you can click on their avatar to go to their current location in the editor and begin following them
- [x] Given a participant that is viewing a different editor than the one you're viewing, verify that you can click on their avatar to go to their current location in their current editor and begin following them
- [x] Given a participant that is viewing a non-portal pane item (e.g., the Atom "Settings" tab), verify that you cannot click on their avatar to follow them
- [x] When you are following another participant, verify that:
    - [x] Scrolling to a different area of the editor disconnects your tether
    - [x] Switching to another editor disconnects your tether
    - [x] When they scroll to an area of the editor that isn't visible in your viewport, your viewport scrolls to their new position
    - [x] When they switch to another editor in the portal, you follow them to that new editor
    - [x] When they switch focus to a non-portal pane item:
        - [x] You see their avatar's opacity lighten, but you remain at your current position in your current editor
        - [x] When they switch focus to an editor in the portal, you see their avatar's opacity darken, and you resume following them

### TODO

- [x] Render all site position avatars in the bottom right of the portal pane item
- [x] If user is viewing a non-portal pane item, adjust opacity of avatar (e.g., see right-most avatar in this [mockup](https://user-images.githubusercontent.com/378023/34096283-20a627cc-e418-11e7-8041-7c5b682ddae0.png))
- [x] Gracefully handle a large number of participants (e.g., https://github.com/atom/teletype/issues/286#issuecomment-352648977)
- [x] Define a *Verification Process* above and execute it
- [x] Merge https://github.com/atom/teletype-client/pull/54, publish a new release of teletype-client, and update teletype to depend on that new version
- ~~Add padding to avoid covering the last few lines~~ UPDATE: See https://github.com/atom/teletype/pull/332#issuecomment-371046601

### Applicable Issues

Closes #268.
Closes #286.

### CODE REVIEW

File: styles/teletype.less [hunk 1]
```diff
@@ -365,48 +367,13 @@
...
-    .SitePositionsComponent-site {
-      margin-left: @avatar-margin;
-      flex-direction: column;
-      justify-content: flex-end;
-      align-items: center;
+    // TODO Improve styling of link icon
+    .icon-link {
+      position: absolute;
+      padding-left: 28px;
+      padding-top: 28px;
```
Comment (jasonrudolph) [inline styles/teletype.less]: @simurai: I really dig the what the link icon appears in your mockup:



This pull request currently hacks in a pretty rudimentary implementation: 

image

Would you be willing to update the styling to match your mockup? You can feel free to push directly to this branch.

Comment (simurai) [reply styles/teletype.less]: Added an extra pseudo element as cover:

teletype

**Comment by simurai** [comment]
> - [ ] Add padding to avoid covering the last few lines

This seems harder than it sounds. I tried adding `padding-bottom: 70px` to some elements, but most things in the editor are absolute positioned. I think it needs the same mechanic like the "Scroll Past End" setting. That one increases the `height` of `.lines` and parent container. So if there isn't an easy solution, I would say we can punt on this for now. Like in case somebody runs into the problem that the avatars cover some code, they can:

- Just add a few extra lines
- Resize the window
- Enable  "Scroll Past End"

Not the best experience, but also not unsolvable and there are other higher priorities.

**Comment by simurai** [comment]
> - [ ] Gracefully handle a large number of participants (e.g., https://github.com/atom/teletype/issues/286#issuecomment-352648977)

Any thoughts on the gracefulness? Currently participants just get cut off when there is not enough space:

teletype

Maybe already fine? Otherwise we would start optimizing for the `0.1%`, if [this chart](https://github.com/atom/teletype/issues/286#issuecomment-352841265) is still accurate.

**Comment by jasonrudolph** [comment]
@simurai: Thanks for these updates! I asked for help with one thing, and you took the time to help with 3 things. :heart_eyes_cat:

One follow-up request: In my testing, the link icon seems to render well when the using a font size of 14 or smaller. When I increase the font above 14, the link icon's position gets distorted. (See examples below.) Can you push up some changes to fix this issue or offer some tips for how I can go about fixing it?

#### 14-point font
screen shot 2018-03-07 at 09 32 48 am
#### 15-point font
screen shot 2018-03-07 at 09 33 30 am
#### 20-point font
screen shot 2018-03-07 at 09 33 38 am

**Comment by jasonrudolph** [comment]
I originally included this as something we might do in this pull request:

> - [ ] Maybe: Show [tooltip](https://github.com/atom/teletype/issues/286#issuecomment-352135030) when hovering over an avatar (e.g., "Follow @as-cii to foobar.cpp:34")

I still like the idea, but this pull request already involves quite a few changes, and this ☝️ would make sense as a standalone enhancement that we address in a future pull request. So, I've removed this task from the list of TODOs in the pull request body.

**Comment by jasonrudolph** [comment]
> Good catch. Ok, it's now locked to a fixed font size...

Thanks, @simurai. :zap::bow:💟

**Comment by xmsid** [comment]
The "Link" icon that appears next to a followed collaborator looks weird on my side
image
I tried different UI/Syntax themes but it still looks like this. is this intended to look like that?

**Comment by 50Wliu** [comment]
@xmsid I think what you're seeing was fixed in #351 so it should be in a future Teletype release.

**Comment by xmsid** [comment]
@50Wliu yeah, I checked that issue it's the same as mine, I'm really looking for more future releases :).

**Comment by jasonrudolph** [comment]
> I'm really looking for more future releases :).

@xmsid: Awesome. Stay tuned. :smile:

**Comment by rsanheim** [comment]
Awesome work y'all! ✨

**Pull request merged**